File: implementations/scala/core/src/main/scala/prism/server/RequestRouter.scala
package prism.server

import cats.effect.IO
import cats.syntax.all._
import prism.core.DeltaComputer
import prism.core.Types._

/** Routes requests and performs smart hydration of object references.
  *
  * The RequestRouter analyzes client state to determine the most efficient way to send object data:
  *   - Full object: Client doesn't have the object or version is too old
  *   - Delta: Client has an older version and delta is more efficient than full object
  *   - Cached: Client already has the current version
  *
  * Supports recursive hydration of nested object references up to a configurable depth.
  *
  * @param objectManager
  *   ObjectManager for accessing objects and client state
  */
class RequestRouter(val objectManager: ObjectManager) {

  /** Hydrate a single object reference based on client state.
    *
    * Analyzes what the client knows and returns the most efficient representation:
    *   - If client has current version: mark as cached
    *   - If client has older version and delta is efficient: return delta
    *   - Otherwise: return full object data
    *
    * @param clientId
    *   Client identifier
    * @param ref
    *   Object reference to hydrate
    * @return
    *   Hydrated reference with appropriate data/delta/cached marker
    */
  def hydrateReference(clientId: String, ref: ObjectReference): IO[HydratedReference] = {
    for {
      clientState <- objectManager.getClientState(clientId)
      currentObj <- objectManager.getObject(ref.id)

      result <- currentObj match {
        case None =>
          // Object doesn't exist
          IO.pure(HydratedReference(id = ref.id, version = 0, cached = false))

        case Some(obj) =>
          val subscription = clientState.getSubscription(ref.id)
          // Use filter from reference, or from subscription, or default
          val filterType = ref.filterType.orElse(subscription.flatMap(s => Some(s.filterType).filter(_ != "default")))
          val filterParams = subscription.flatMap(_.filterParams)

          for {
            filtered <- objectManager.applyFilter(obj, filterType, filterParams)
            clientVersion = clientState.getVersion(ref.id)

            hydrated <- clientVersion match {
              case Some(cv) if cv == filtered.version =>
                // Client has current version - cached
                IO.pure(HydratedReference(
                  id = filtered.id,
                  version = filtered.version,
                  cached = true
                ))

              case Some(cv) if cv < filtered.version =>
                // Client has older version - try delta
                for {
                  delta <- objectManager.getDelta(ref.id, cv, filtered.version, filterType)
                  result <- delta match {
                    case Some(d) if DeltaComputer.isDeltaEfficient(d, filtered) =>
                      // Delta is efficient
                      IO.pure(HydratedReference(
                        id = filtered.id,
                        version = filtered.version,
                        delta = Some(d)
                      ))
                    case _ =>
                      // Delta not efficient or not available - send full object
                      IO.pure(HydratedReference(
                        id = filtered.id,
                        version = filtered.version,
                        data = Some(filtered.data)
                      ))
                  }
                } yield result

              case _ =>
                // Client doesn't have this object - send full
                IO.pure(HydratedReference(
                  id = filtered.id,
                  version = filtered.version,
                  data = Some(filtered.data)
                ))
            }
          } yield hydrated
      }
    } yield result
  }

  /** Hydrate multiple object references.
    *
    * @param clientId
    *   Client identifier
    * @param refs
    *   List of object references to hydrate
    * @return
    *   List of hydrated references
    */
  def hydrateReferences(clientId: String, refs: List[ObjectReference]): IO[List[HydratedReference]] = {
    refs.traverse(ref => hydrateReference(clientId, ref))
  }

  /** Process a request with object references and optional auto-hydration.
    *
    * This is the main entry point for handling client requests. It:
    *   1. Hydrates the primary object references
    *   2. Optionally scans for nested ObjectReferences in the data
    *   3. Recursively hydrates nested references up to maxDepth
    *
    * @param clientId
    *   Client identifier
    * @param refs
    *   Primary object references from the request
    * @param options
    *   Request options (hydration settings)
    * @param currentDepth
    *   Current recursion depth (internal)
    * @return
    *   List of all hydrated references (primary + nested if enabled)
    */
  def processRequest(
      clientId: String,
      refs: List[ObjectReference],
      options: RequestOptions = RequestOptions.default,
      currentDepth: Int = 0
  ): IO[List[HydratedReference]] = {
    if (currentDepth > options.maxHydrationDepth) {
      // Exceeded max depth - return empty
      IO.pure(List.empty)
    } else {
      for {
        // Hydrate primary references
        hydrated <- hydrateReferences(clientId, refs)

        // If auto-hydration enabled, find nested references
        nested <- if (options.hydrateRefs && currentDepth < options.maxHydrationDepth) {
          val nestedRefs = extractNestedReferences(hydrated, options)
          if (nestedRefs.nonEmpty) {
            processRequest(clientId, nestedRefs, options, currentDepth + 1)
          } else {
            IO.pure(List.empty)
          }
        } else {
          IO.pure(List.empty)
        }

        // Optionally subscribe to nested references
        _ <- if (options.subscribeToRefs && nested.nonEmpty) {
          nested.traverse { hr =>
            objectManager.subscribe(
              clientId,
              hr.id,
              options.filterType.getOrElse("default"),
              None,
              ongoing = true
            )
          }.void
        } else {
          IO.unit
        }
      } yield hydrated ++ nested
    }
  }

  /** Extract nested ObjectReferences from hydrated data.
    *
    * Scans JSON data for objects that match the ObjectReference schema and extracts them for hydration.
    *
    * @param hydrated
    *   Hydrated references to scan
    * @param options
    *   Request options (default filter to apply)
    * @return
    *   List of nested object references found
    */
  private def extractNestedReferences(
      hydrated: List[HydratedReference],
      options: RequestOptions
  ): List[ObjectReference] = {
    hydrated.flatMap { hr =>
      hr.data match {
        case Some(data) => scanForReferences(data, options.filterType.getOrElse("default"))
        case None       => List.empty
      }
    }
  }

  /** Recursively scan JSON for ObjectReference-like objects.
    *
    * Looks for objects with "id" field and optionally "version", "filter", etc.
    *
    * @param value
    *   JSON value to scan
    * @param defaultFilter
    *   Default filter to apply to found references
    * @return
    *   List of object references found
    */
  private def scanForReferences(value: ujson.Value, defaultFilter: String): List[ObjectReference] = {
    value match {
      case obj: ujson.Obj =>
        // Check if this object is itself a reference
        val selfRef = if (obj.value.contains("id")) {
          val id = obj("id").strOpt.getOrElse("")
          if (id.nonEmpty) {
            val version = obj.value.get("version").flatMap(_.numOpt.map(_.toInt)).getOrElse(0)
            val filterType = obj.value.get("filterType").flatMap(_.strOpt).orElse(
              obj.value.get("filter").flatMap(_.strOpt)
            )
            val subscribe = obj.value.get("subscribe").flatMap(_.boolOpt).getOrElse(false)

            Some(ObjectReference(
              id = id,
              version = version,
              filterType = filterType,
              subscribe = subscribe
            ))
          } else {
            None
          }
        } else {
          None
        }

        // Recursively scan nested values
        val nestedRefs = obj.value.values.flatMap(v => scanForReferences(v, defaultFilter)).toList

        selfRef.toList ++ nestedRefs

      case arr: ujson.Arr =>
        // Scan array elements
        arr.value.flatMap(v => scanForReferences(v, defaultFilter)).toList

      case _ =>
        // Primitive value - no references
        List.empty
    }
  }

  /** Create temporary subscriptions for a one-time request.
    *
    * Subscribes to all referenced objects as temporary (will be cleared after response).
    *
    * @param clientId
    *   Client identifier
    * @param refs
    *   Object references to subscribe to
    * @param filterType
    *   Default filter to apply
    * @return
    *   Unit
    */
  def createTemporarySubscriptions(
      clientId: String,
      refs: List[ObjectReference],
      filterType: String = "default"
  ): IO[Unit] = {
    refs.traverse { ref =>
      objectManager.subscribe(
        clientId,
        ref.id,
        ref.filterType.getOrElse(filterType),
        None, // Filter params come from subscription
        ongoing = false // Temporary subscription
      )
    }.void
  }
}

object RequestRouter {

  /** Create a new RequestRouter.
    *
    * @param objectManager
    *   ObjectManager instance
    * @return
    *   A new RequestRouter
    */
  def create(objectManager: ObjectManager): RequestRouter = {
    new RequestRouter(objectManager)
  }
}

File: implementations/python/prism/server/request_router.py
"""Request router and response enhancer for smart object hydration."""

from abc import ABC, abstractmethod
from typing import Any

from prism.core.delta import DeltaComputer
from prism.core.protocol import RequestMessage, ResponseMessage
from prism.core.types import (
    Delta,
    HydratedReference,
    ObjectReference,
    RequestOptions,
)
from prism.server.object_manager import PrismObjectManager


class BusinessHandler(ABC):
    """Abstract interface for business logic handlers.

    Business logic remains independent of Prism - handlers just process
    requests and return data (potentially with ObjectReferences).
    """

    @abstractmethod
    async def process(self, request_type: str, payload: dict[str, Any]) -> Any:
        """Process a business logic request.

        Args:
            request_type: Type of request
            payload: Request payload

        Returns:
            Response data (may contain ObjectReferences)

        Raises:
            Exception: If request processing fails
        """
        pass


class RequestRouter:
    """Routes requests to business logic and enhances responses with smart hydration."""

    def __init__(self, object_manager: PrismObjectManager, business_handler: BusinessHandler):
        """Initialize request router.

        Args:
            object_manager: Prism object manager
            business_handler: Business logic handler
        """
        self.object_manager = object_manager
        self.business_handler = business_handler
        self.delta_computer = DeltaComputer()

    async def handle_request(self, client_id: str, request: RequestMessage) -> ResponseMessage:
        """Handle a request with smart response hydration.

        Args:
            client_id: Client identifier
            request: The request message

        Returns:
            Enhanced response message
        """
        try:
            # Parse options
            options = RequestOptions(**(request.options or {}))

            # Process business logic
            result = await self.business_handler.process(request.request_type, request.payload)

            # Enhance response with smart hydration
            hydrated = []
            if options.hydrate_refs:
                hydrated = await self.enhance_response(client_id, result, options)

            return ResponseMessage(
                request_id=request.request_id,
                success=True,
                data=result,
                hydrated=hydrated,
            )

        except Exception as e:
            return ResponseMessage(
                request_id=request.request_id,
                success=False,
                error=str(e),
            )

    async def enhance_response(
        self, client_id: str, data: Any, options: RequestOptions
    ) -> list[HydratedReference]:
        """Scan response for object references and hydrate based on client state.

        Args:
            client_id: Client identifier
            data: Response data to scan
            options: Request options

        Returns:
            List of hydrated references
        """
        # Extract all object references from the response
        references = self._extract_references(data)

        if not references:
            return []

        # Hydrate each reference based on client state
        hydrated = []
        client_state = self.object_manager.get_client_state(client_id)

        for ref in references:
            # Get current object
            obj = await self.object_manager.storage.get_current(ref.id)
            if obj is None:
                continue

            # Apply filter if specified
            filter_type = ref.filter_type or options.filter_type or "default"
            if filter_type != "default":
                obj = self.object_manager.filters.apply(obj, filter_type)

            # Determine what to send based on client state
            known_version = client_state.get_version(ref.id)

            if known_version is None:
                # Client doesn't have object - send full
                hydrated.append(HydratedReference(id=ref.id, version=ref.version, data=obj.data))

                # Auto-subscribe if requested
                if options.subscribe_to_refs or ref.subscribe:
                    from prism.core.protocol import SubscribeMessage

                    await self.object_manager.handle_subscribe(
                        client_id,
                        SubscribeMessage(
                            object_id=ref.id,
                            filter_type=filter_type,
                        ),
                    )

            elif known_version < ref.version:
                # Client has older version - try delta
                delta = await self._compute_delta_for_ref(
                    ref.id, known_version, ref.version, filter_type
                )

                if delta and self.delta_computer.is_delta_efficient(delta, obj):
                    # Send delta
                    hydrated.append(HydratedReference(id=ref.id, version=ref.version, delta=delta))
                else:
                    # Delta too large - send full
                    hydrated.append(
                        HydratedReference(id=ref.id, version=ref.version, data=obj.data)
                    )

                # Update client's known version
                client_state.update_version(ref.id, ref.version)

            else:
                # Client has current version - mark as cached
                hydrated.append(HydratedReference(id=ref.id, version=ref.version, cached=True))

        return hydrated

    def _extract_references(self, data: Any, max_depth: int = 5) -> list[ObjectReference]:
        """Extract all ObjectReferences from response data.

        Args:
            data: Data to scan (dict, list, or primitive)
            max_depth: Maximum recursion depth

        Returns:
            List of found ObjectReferences
        """
        if max_depth <= 0:
            return []

        references: list[ObjectReference] = []

        # Check if this is already an ObjectReference instance
        if isinstance(data, ObjectReference):
            references.append(data)
            return references

        if isinstance(data, dict):
            # Check if this dict is an ObjectReference
            if self._is_object_reference(data):
                try:
                    ref = ObjectReference(**data)
                    references.append(ref)
                except Exception:
                    pass
            else:
                # Recurse into dict values
                for value in data.values():
                    references.extend(self._extract_references(value, max_depth - 1))

        elif isinstance(data, list):
            # Recurse into list items
            for item in data:
                references.extend(self._extract_references(item, max_depth - 1))

        return references

    def _is_object_reference(self, data: dict[str, Any]) -> bool:
        """Check if a dict represents an ObjectReference.

        Args:
            data: Dict to check

        Returns:
            True if it looks like an ObjectReference
        """
        # ObjectReference must have 'id' and 'version'
        return "id" in data and "version" in data and isinstance(data.get("version"), int)

    async def _compute_delta_for_ref(
        self, object_id: str, from_version: int, to_version: int, filter_type: str
    ) -> Delta | None:
        """Compute delta between filtered versions.

        Args:
            object_id: Object ID
            from_version: Starting version
            to_version: Target version
            filter_type: Filter to apply

        Returns:
            Delta or None if unavailable
        """
        try:
            from_obj = await self.object_manager.storage.get_version(object_id, from_version)
            to_obj = await self.object_manager.storage.get_version(object_id, to_version)

            if from_obj is None or to_obj is None:
                return None

            # Apply filters to both versions
            if filter_type != "default":
                from_obj = self.object_manager.filters.apply(from_obj, filter_type)
                to_obj = self.object_manager.filters.apply(to_obj, filter_type)

            return self.delta_computer.compute_delta(from_obj, to_obj)

        except Exception:
            return None
